src/library.js

```diff
@@ -3781,6 +3781,106 @@ LibraryManager.library = {
       }
     }
   },
+
+  // Callable in pthread without __proxy needed.
+  emscripten_exit_with_live_runtime__sig: 'v',
+  emscripten_exit_with_live_runtime__deps: ['emscripten_runtime_keepalive_push'],
+  emscripten_exit_with_live_runtime: function() {
+    _emscripten_runtime_keepalive_push();
+    throw 'unwind';
+  },
+
+  emscripten_force_exit__deps: ['$runtimeKeepaliveCounter'],
+  emscripten_force_exit__proxy: 'sync',
+  emscripten_force_exit__sig: 'vi',
+  emscripten_force_exit: function(status) {
+#if EXIT_RUNTIME == 0
+#if ASSERTIONS
+    warnOnce('emscripten_force_exit cannot actually shut down the runtime, as the build does not have EXIT_RUNTIME set');
+#endif
+#endif
+    noExitRuntime = false;
+    runtimeKeepaliveCounter = 0;
+    exit(status);
+  },
+
+  $runtimeKeepaliveCounter: 0,
+
+  $keepRuntimeAlive__deps: ['$runtimeKeepaliveCounter'],
+  $keepRuntimeAlive: function() {
+    return noExitRuntime || runtimeKeepaliveCounter > 0;
+  },
+
+  // Callable in pthread without __proxy needed.
+  emscripten_runtime_keepalive_push__sig: 'v',
+  emscripten_runtime_keepalive_push__deps: ['$runtimeKeepaliveCounter'],
+  emscripten_runtime_keepalive_push: function() {
+    runtimeKeepaliveCounter += 1;
```

Comment:
`++runtimeKeepaliveCounter;` for one byte code win (sorry, my eyes can't unsee things like these after working on 16k Wasm sizedemos)

Push/pop implies stack semantics, but this is not a stack, but a refcount. I.e. via static reading, one might think that with stack semantics, if A pushes and B pushes, then the only valid behavior is for B to pop first before A, but that is certainly not the case here. Maybe a better name would be to use `_addref` and `_decref` to imply reference counting?

Though I wonder if a refcount counter should be used at all for this, since it can be extremely hard to debug when the refcounts don't match. Rather, maybe an API `emscripten_runtime_keepalive_add_ref('stringIdentifier');` and `emscripten_runtime_keepalive_remove_ref('stringIdentifier');` would make more sense for debuggability in mind? i.e. the similar abstraction as pre run dependencies have (if code size is a concern, then in release builds, those could fall back to refcounting and ignore the string value, and closure should strip out the string literals from caller side)